sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala

```diff
@@ -137,6 +157,163 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
   protected def doExecute(): RDD[InternalRow]
 
   /**
+    * Whether this SparkPlan support whole stage codegen or not.
+    */
+  protected def supportCodeGen: Boolean = false
+
+  class CompileFailure(e: Exception) extends Exception {}
+
+  /**
+    * Returns an RDD of InternalRow that using generated code to process them.
+    *
+    * Here is the call graph of three SparkPlan, A and B support codegen, but C does not.
+    *
+    *   SparkPlan A            SparkPlan B          SparkPlan C
+    * ===============================================================
+    * -> doExecute()
+    *     |
+    *  doCodeGen()
+    *     |
+    *  produce()
+    *     |
+    *  doProduce() -------->   produce()
```

Comment:
Can you comment produce/consume more? What is the state of the codegen context when it is called and where does it leave it?